design-docs/012-python-namespaces

```diff
@@ -0,0 +1,216 @@
+  - `conflict.Conflict`
+  - `conflict.ConflictDetector`
+  - `conflict.ConflictSolver`
+  - `conflict.ConflictType`
+  - `conflict.VersionSelection`
+- `exceptions`
+  - `exceptions.ConflictError`
+  - `exceptions.IcechunkError`
+  - `exceptions.RebaseFailedError`
+- `logging`
+  - `logging.initialize_logs`,
+  - `logging.set_logs_filter`,
+- `manifests`
+  - `manifests.ManifestConfig`
+  - `manifests.ManifestFileInfo`
+  - `manifests.ManifestPreloadCondition`
+  - `manifests.ManifestPreloadConfig`
+  - `manifests.ManifestSplitCondition`
+  - `manifests.ManifestSplitDimCondition`
+  - `manifests.ManifestSplittingConfig`
+- `snapshots`
+  - `snapshots.SnapshotInfo`
+  - `snapshots.Diff`
+- `virtual`
+  - `virtual.VirtualChunkContainer`
+  - `virtual.VirtualChunkSpec`
+
+## Outstanding questions
+
+- It's generally a bit confusing that there are multiple ways to make most classes, e.g. `local_filesystem_store` and `Storage.new_local_filesystem`. Do we really need both?
+  - If we only kept `Storage` we could get rid of a _lot_ of functions in favour of methods...
+  - Alternatively should base classes that aren't really meant to be called by the user (e.g. `Storage`) be public API at all?
+- What's the difference between `storage.local_filesystem_storage` and `storage.local_filesystem_store`?
+  - Similarly how can we better disambiguate `storage.s3_storage` and `storage.s3_store`?
+- Is there a similar question for credentials?
+- Should cloud-provider-specific classes be further grouped together within `.storage`/`.credentials`?
+- Should all configs live together?
+  - Should `StorageSettings` really be a `StorageConfig`? Should it live with configs or in `storage`?
+  - Similarly should `S3Options` really be `S3Config`?
+- Should summary dataclasses such as `Diff` and `GCSummary` live in the top-level?
+- Is `distributed` meant to be a public module at all? Why is it's API defined in a different place? Should it be merged with `dask`?
+- Union types like `AnyCredential` are re-exported as top-level API for importing, but it seems like they are only useful for internal type-hinting purposes?
+
+## Implementation steps (including deprecation cycle)
+
+We need to implement this in multiple steps, including a deprecation cycle which maintains backwards-compatibility until the cycle is complete.
+
+1. **Decide on public API structure (i.e. review then merge this design doc)**
+
+2. **Move everything into desired locations whilst keeping all top-level imports**
+
+We should add tests that both import paths work.
+
+3. **Change how docs render to display using fine-grained namespaces instead of single top-level namespace**
+
+We can probably configure the API docs to display the submodules first and ignore the top-level imports, even whilst the top-level imports are still available.
+
+4. **Add deprecation warning pointing users to specific fine-grained import**
+
+User experience should look like this:
+
+```python
+# works, but emits a warning
```

Comment:
I think we should use Icechunk 2 as the moment to make the change and start warning people about deprecation.